torchx/c_src/torchx.cpp

```diff
@@ -599,6 +599,15 @@ NIF(sort)
   TENSOR(std::get<0>(result));
 }
 
+NIF(clip)
+{
+  TENSOR_PARAM(0, t);
+  SCALAR_PARAM(1, min);
+  SCALAR_PARAM(2, max);
+
+  TENSOR(torch::clip(*t, min, max));
```

Comment:
This NIF doesn't work for mixed types for some reason. For example, the output of using integers is:

```ex
iex> Nx.clip(Nx.tensor([[1, 2, 3], [4, 5, 6]], names: [:x, :y]), Nx.tensor(2), Nx.tensor(4))
#Nx.Tensor<
  s64[x: 2][y: 3]
  Torchx.Backend(cpu)
  [
    [2, 2, 3],
    [4, 4, 4]
  ]
>
```
But, when we use floating type numbers(scalars or tensors)

```ex
iex> Nx.clip(Nx.tensor([[1, 2, 3], [4, 5, 6]], names: [:x, :y]), Nx.tensor(2.1), Nx.tensor(4))
#Nx.Tensor<
  f32[x: 2][y: 3]
  Torchx.Backend(cpu)
  [
    [2.802596928649634e-45, 0.0, 2.802596928649634e-45],
    [0.0, 4.203895392974451e-45, 0.0]
  ]
>
```

I noticed that the `min` and `max` variables are passed as tensors from the `backend.ex` file, but when I replace `SCALAR_PARAM` with `TENSOR_PARAM` I get the error:

```ex
iex(1)>   Nx.clip(Nx.tensor([[1, 2, 3], [4, 5, 6]], names: [:x, :y]), Nx.tensor(2), Nx.tensor(4))
** (RuntimeError) Torchx: Unable to get min tensor param in NIF.clip/3
    (torchx 0.1.0-dev) lib/torchx.ex:328: Torchx.unwrap!/1
    (torchx 0.1.0-dev) lib/torchx.ex:331: Torchx.unwrap_tensor!/2
    (torchx 0.1.0-dev) lib/torchx/backend.ex:763: Torchx.Backend.clip/4
```

This errors occurs both when I use `to_number` and without it in the `backend.ex` file. Could someone please help me fix this issue? 😄

Reply:
If you replace `SCALAR_PARAM` with `TENSOR_PARAM` in the `.cpp` file, you need to remove the `to_number` call in the  backend.ex file. Also, you need to rename the params in `deftensor clip(tensor, tensor_min, tensor_max)`, to ensure they contain tensors.

However, I suggest keeping the `to_number` calls as the code is now. You'll probably need to promote the args to float whenever one of them is a float already, which should get rid of the type mismatch issue

Reply:
Oh, I just read the "floating" result (on mobile it didn't show up). What's probably happening is that the `4` is being passed as an integer but is being bitcast as a float due to `2.1` coming first.

Reply:
Looking at the docs it seems that `torch::clip` does accept tensors for `min` and `max`, so perhaps you could try that? Just change the `deftensor` as I pointed out and change `to_number` to `from_nx` in Torchx.Backend

Reply:
Thanks for your review @polvalente! Indeed I did try to use tensors directly and use `TENSOR_PARAM`, and I remember seeing the following trace:

```ex
** (SyntaxError) iex:3:18: unexpected ( after alias Backend. Function names and identifiers in Elixir start with lowercase characters or underscore. For example:

    hello_world()
    _starting_with_underscore()
    numb3rs_are_allowed()
    may_finish_with_question_mark?()
    may_finish_with_exclamation_mark!()

Unexpected token: (
``` 

I think I missed the `from_nx` change for the min and max tensors. It does work as expected after applying the changes you've suggested! 😄